Improvements to funding and distribution forms

johnnyd-eth opened the pull request:
## What does this PR do and why?

This PR improves how the funding target is set/changed, as well as how individual distributions are set/changed:

1. Allows user to input **funding target after Juicebox fee** and have it set the **funding target**
2. Allows user to set distribution payout as an **amount** rather than just a **percentage** of the funding target

These were ideas sent to me from @mejango 

## Screenshots or screen recordings

Overview:

https://user-images.githubusercontent.com/96150256/147845351-9d85070c-81a8-494e-b45d-fd5019bbdd49.mp4

Rounds **amount** to floor of 2 d.p. **percentage**:


https://user-images.githubusercontent.com/96150256/147865936-1341550a-464b-4456-a3a2-3ab5d44e6651.mp4

Fix bug where subsequent payout of a project payout gets assigned that project (before/after):

https://user-images.githubusercontent.com/96150256/147891846-af99cd86-35ad-477c-b6dc-778c73e9c2a2.mp4

Address @peripheralist's 5 points of feedback:

https://user-images.githubusercontent.com/96150256/148140402-f241a3c4-8caf-4fc3-a70d-f7f9555ac2d2.mp4


## Acceptance checklist

- [x] I have evaluted the [Approval Guidelines](../../CONTRIBUTING.md#approval-guidelines) for this PR.

johnnyd-eth commented on src/utils/math.ts:
```diff
@@ -37,5 +37,5 @@ export const amountSubFee = (amount?: BigNumber, feePercent?: BigNumber) => {
 export const amountAddFee = (amount?: BigNumber, feePercent?: BigNumber) => {
   if (!feePercent || !amount) return
-  return amount.add(feeForAmount(amount, feePercent) ?? 0)
+  return amount.add(amount.mul(100).div(feePercent.mul(200)))
```
Previously, amountAddFee was not the inverse of amountSubFee, i.e. amount != amountSubFee(amountAddFee(amount)).

johnnyd-eth commented on src/components/shared/formItems/formHelpers.tsx:
```diff
@@ -0,0 +1,93 @@
+  return parseInt(
+    stripCommas(
+      formatWad(
+        amountSubFee(parseWad(stripCommas(target)), fee)
```
This is horrible, I know. But .toNumber() was giving me so much grief I just gave up and went with this

peripheralist replied:
it looks like `getAmountFromPercent` could just return a string instead of a number, so we could simplify `parseInt(stripCommas(formatWad(...` into just `fromWad(`

johnnyd-eth commented:
@mejango second vid I've attached in the description shows the amount rounding to the floor of the percentage to 2 decimal places. Is this the behavior you were expecting?

mejango commented:
@johnnyd-eth nice! Might just wanna let the user know on the form itself, maybe some subtext "will be rounded up to X%"

johnnyd-eth commented:
> @johnnyd-eth nice! Might just wanna let the user know on the form itself, maybe some subtext "will be rounded up to X%"

@johnnyd-eth Yep good idea. I've added that and updated the second video which shows the message when the user inputs an amount that will be rounded.

aeolianeth commented on src/components/Create/BudgetForm.tsx:
```diff
@@ -93,7 +107,11 @@ export default function BudgetForm({
               checked={showFundingFields}
               onChange={checked => {
-                setTarget(checked ? '10000' : maxIntStr || '0')
+                let target = checked ? '10000' : maxIntStr || '0'
```
**nitpick (non-blocking)**: think we can make this a const:

```suggestion
                const target = checked ? '10000' : maxIntStr || '0'
```

Also, WDYT about making `'10000'` a constant at the top of the file, something like
```js
const DEFAULT_TARGET = '10000';
```
```

aeolianeth commented on src/components/Create/BudgetForm.tsx:
```diff
@@ -116,8 +134,23 @@ export default function BudgetForm({
+            target={target}
+            targetSubFee={targetSubFee}
+            onTargetChange={target => {
+              setTarget(target || '0')
```
**suggestion (non-blocking)**: we could be a little more strict here (and in other instances below) and use the nullish coalescing operator:

```suggestion
              setTarget(target ?? '0')
```

aeolianeth commented on src/components/shared/formItems/formHelpers.tsx:
```diff
@@ -0,0 +1,107 @@
+}
+
+// Ensures value is greater than 0 and less than 100
+export function validatePercentage(percent: number | undefined) {
```
**praise**: nice helper!

aeolianeth commented on src/components/shared/formItems/formHelpers.tsx:
```diff
@@ -0,0 +1,107 @@
+export function validateEthAddress(
+  address: string,
+  mods: PayoutMod[],
+  modalMode: string | undefined,
```
**question**: I seem to remember we have a `modalMode` enum type somewhere. Can we reuse it here?

aeolianeth commented on src/components/shared/formItems/formHelpers.tsx:
```diff
@@ -0,0 +1,107 @@
+  target: string,
+  fee: BigNumber | undefined,
+) => {
+  let newTargetSubFee = amountSubFee(parseWad(target ?? ''), fee)
```
**nitpick (non-blocking)**: can be `const`

aeolianeth commented on src/utils/formatNumber.ts:
```diff
@@ -33,6 +33,11 @@ export const formatWad = (amt?: BigNumberish, config?: FormatConfig) => {
 }
 
+// Strips string of all commas
+export const stripCommas = (string: string) => {
```
**praise**: love this

peripheralist replied:
we may be able to get around needing this if we use `fromWad()` instead of `formatWad()`!

aeolianeth commented on src/components/shared/formItems/formHelpers.tsx:
```diff
@@ -0,0 +1,107 @@
+}
+
+// Returns number of decimal places in a given number
+export function countDecimals(value: number | undefined) {
```
**suggestion (non-blocking)**: perhaps `countDecimalPlaces` for the name?

aeolianeth commented on src/components/shared/formItems/formHelpers.tsx:
```diff
@@ -0,0 +1,107 @@
+  target: string,
+  fee: BigNumber | undefined,
+) {
+  let targetSubFeeBN = amountSubFee(parseWad(stripCommas(target)), fee)
```
**suggestion (non-blocking)**: can `const` these bad boys too

aeolianeth reviewed (approved):
Hey @johnnyd-eth killer job here, love seeing these form improvements. I left a few minor comments here and there for you, but happy to approve ✅ 

![lol](https://media.giphy.com/media/KIxBn0rcMbhII/giphy.gif)

aeolianeth commented on src/components/shared/formItems/ProjectPayoutMods.tsx:
```diff
@@ -461,35 +483,59 @@ export default function ProjectPayoutMods({
+              }}
+            >
+              <div style={{ marginRight: 5 }}>
+                <CurrencySymbol currency={currency} />
```
**suggestion**: antd has a nice way of handling this with the `addonBefore` prop on the `Input` component. 
Docs are here (look for `Pre/Post tab`): https://ant.design/components/input/

aeolianeth commented:
@johnnyd-eth 1 UX comment also: in the funding target second input, should we add the ETH/USD currency badge and sync it with the main input?

johnnyd-eth commented on src/components/shared/formItems/ProjectPayoutMods.tsx:
```diff
@@ -329,10 +329,10 @@ export default function ProjectPayoutMods({
-      setSettingHandle(handle)
-      setSettingHandleIndex(editingModIndex)
-    }
+    // if (handle) {
```
@mejango this fixes the duplicate project bug you showed me at 3:47 in your video. 

@peripheralist the third video I've attached shows the problem and the fix as a result of removing this block of code that sets the project handle in the state after the mod has been submitted. Do you know if there's any other reason it should be here?

johnnyd-eth commented:
> @johnnyd-eth 1 UX comment also: in the funding target second input, should we add the ETH/USD currency badge and sync it with the main input?

Good idea, I've added this:

![Kapture 2022-01-03 at 11 11 41](https://user-images.githubusercontent.com/96150256/147894103-fb6d9438-95ee-4c0a-8f89-4be06ea3cf56.gif)

johnnyd-eth commented:
Thanks for your great feedback @aeolianeth. I've addressed all your points other than using the `addonBefore` prop which will entail extra styling work which I didn't think was worth.

peripheralist commented:
this is huge @johnnyd-eth! awesome work. i love the idea of just using two inputs for this

it looks like a few comments were resolved without changes being made (the `addOnBefore` and turning `let`s into `const`s. did you forget to push your latest? 😏)

i also added a couple more comments about number formatting

---

some design thoughts:

1) could the "after fee" input be expanded like this to offer some more room? 
![image](https://user-images.githubusercontent.com/79433522/147955867-a00cbb6c-b60b-4137-8aa6-d9bf733c2981.png)

2) could the $10,000 default value be used to set the _after fee_ amount instead of the pre-fee amount, so the after-fee is a clean number?
![image](https://user-images.githubusercontent.com/79433522/147956816-1c87c4dc-d969-4780-a40d-220781d2a2eb.png)

3) this "will be rounded" message is awesome, though maybe it'd make more sense to show the actual $/ETH amount the payout will be rounded to, instead of the percent which is shown right below? 
![image](https://user-images.githubusercontent.com/79433522/147956204-e85ccb0e-f5d7-4474-846a-13e4b9d99d16.png)

4) it would be helpful to have an error message next to a payout amount when a number larger than the target is entered, instead of just showing the error state on the percent input 
![image](https://user-images.githubusercontent.com/79433522/147956433-4f0bea2a-15bc-4693-bcfc-dd26f9d63f77.png)

5) it looks like the percent doesn't update when inputting decimals in the payout (should be 40%)
![image](https://user-images.githubusercontent.com/79433522/147956593-0c49a688-ab3c-4316-9237-8122b48fc955.png)

johnnyd-eth commented:
Thanks for your comments @peripheralist all great points! I've addressed them all and attached another video in the description showing the updates.

johnnyd-eth commented:
> this is huge @johnnyd-eth! awesome work. i love the idea of just using two inputs for this
> 
> it looks like a few comments were resolved without changes being made (the `addOnBefore` and turning `let`s into `const`s. did you forget to push your latest? 😏)
> 
> i also added a couple more comments about number formatting
> 
> some design thoughts:
> 
> 1. could the "after fee" input be expanded like this to offer some more room?
>    ![image](https://user-images.githubusercontent.com/79433522/147955867-a00cbb6c-b60b-4137-8aa6-d9bf733c2981.png)
> 2. could the $10,000 default value be used to set the _after fee_ amount instead of the pre-fee amount, so the after-fee is a clean number?
>    ![image](https://user-images.githubusercontent.com/79433522/147956816-1c87c4dc-d969-4780-a40d-220781d2a2eb.png)
> 3. this "will be rounded" message is awesome, though maybe it'd make more sense to show the actual $/ETH amount the payout will be rounded to, instead of the percent which is shown right below?
>    ![image](https://user-images.githubusercontent.com/79433522/147956204-e85ccb0e-f5d7-4474-846a-13e4b9d99d16.png)
> 4. it would be helpful to have an error message next to a payout amount when a number larger than the target is entered, instead of just showing the error state on the percent input
>    ![image](https://user-images.githubusercontent.com/79433522/147956433-4f0bea2a-15bc-4693-bcfc-dd26f9d63f77.png)
> 5. it looks like the percent doesn't update when inputting decimals in the payout (should be 40%)
>    ![image](https://user-images.githubusercontent.com/79433522/147956593-0c49a688-ab3c-4316-9237-8122b48fc955.png)

Thanks for your comments @peripheralist all great points! I've addressed them all and attached another video in the description showing the updates.

johnnyd-eth commented:
Thanks for your comments @peripheralist all great points! I've addressed them all and attached another video in the description showing the updates.

peripheralist commented on src/components/shared/formItems/formHelpers.tsx:
```diff
@@ -0,0 +1,117 @@
+  fee: BigNumber | undefined,
+) => {
+  const newTargetSubFee = amountSubFee(parseWad(target ?? ''), fee)
+  return stripCommas(formatWad(newTargetSubFee, { decimals: 4 }) || '0') // formatWad returns formatted bigNum with commas, must remove
```
we should be able to replace `stripCommas(formatWad(x))` with `fromWad(x)`

johnnyd-eth replied:
Oh yep I've just fixed those up. Just wondering if we're happy showing any number of decimals places on the funding target inputs or if we want to round them at some point:

<img width="630" alt="Screen Shot 2022-01-08 at 10 05 22 am" src="https://user-images.githubusercontent.com/96150256/148622345-8a3c84ea-27f7-4973-974f-354bc3992ec9.png">

johnnyd-eth replied:
@peripheralist Oh yep I missed those! Just wondering if we're happy showing any number of decimals places or if we want to round them at some point (Jango tells me we store funding targets to 16 d.p.'s so it's not inaccurate as it's shown here): 

<img width="630" alt="Screen Shot 2022-01-08 at 10 05 22 am" src="https://user-images.githubusercontent.com/96150256/148622614-cfef2f98-67fc-44b4-a21c-7d2969bd1a37.png">

peripheralist replied:
I'm down to just show the full amount for clarity! I imagine it will usually be more clean if users are inputting round numbers

peripheralist commented on src/components/shared/inputs/FormattedNumberInput.tsx:
```diff
@@ -57,57 +62,64 @@ export default function FormattedNumberInput({
-      </div>
+    <div className="formatted-number-input">
+      {currency ? (
+        <div style={{ marginRight: 5, marginTop: 4 }}>{currency}</div>
```
love that you've added in a currency option here. could we replace it with something like this to be more visually consistent? 
```
<FormattedNumberInput
    ...
    accessory={
        <InputAccessoryButton
              content={currencyName(currency)}
        />}
/>
```

johnnyd-eth replied:
Ahh yep cool forgot about that component, thanks! 

<img width="263" alt="Screen Shot 2022-01-08 at 10 20 59 am" src="https://user-images.githubusercontent.com/96150256/148623243-c98dc235-a458-4342-a620-2e71a385edf7.png">

peripheralist reviewed (changes requested):
i think just a few more little things to touch up before this is ready to go!

peripheralist commented:
another little thing @johnnyd-eth :

the currency doesn't seem to reflect correctly in this error message. this project is in ETH, but the message still shows a "$"
![image](https://user-images.githubusercontent.com/79433522/148663291-b73532e5-5a59-4bbe-ab20-eaf87549f3be.png)

also, could we make that amount input full width as well to match the others?

johnnyd-eth commented:
> another little thing @johnnyd-eth :
> 
> the currency doesn't seem to reflect correctly in this error message. this project is in ETH, but the message still shows a "$" ![image](https://user-images.githubusercontent.com/79433522/148663291-b73532e5-5a59-4bbe-ab20-eaf87549f3be.png)
> 
> also, could we make that amount input full width as well to match the others?
> 
> getting there slowly but surely! thanks for taking so much time with this. will be huge once it's merged in.

Another great pickup @peripheralist thanks! Fixed both of those points up. 
<img width="494" alt="Screen Shot 2022-01-09 at 9 54 06 am" src="https://user-images.githubusercontent.com/96150256/148663879-5046c4b8-6ad2-4487-a749-e3f46acc45d4.png">

peripheralist commented:
@johnnyd-eth did you push changes?

johnnyd-eth commented:
> @johnnyd-eth did you push changes?

Christ! Up now

peripheralist commented:
let's GOOOO! awesome work @johnnyd-eth

johnnyd-eth commented:
> let's GOOOO! awesome work @johnnyd-eth

Yewwww! thanks @peripheralist
